// @nodable/entities/src/entities.js
// ---------------------------------------------------------------------------
// Complete HTML5 named entity reference
// Organized by logical categories for easy maintenance and selective importing
// ---------------------------------------------------------------------------

/**
 * Basic Latin & Special Characters
 * @type {Record<string, string>}
 */
export const BASIC_LATIN = {
  amp: '&',
  AMP: '&',
  lt: '<',
  LT: '<',
  gt: '>',
  GT: '>',
  quot: '"',
  QUOT: '"',
  apos: "'",
  lsquo: 'вҖҳ',
  rsquo: 'вҖҷ',
  ldquo: 'вҖң',
  rdquo: 'вҖқ',
  lsquor: 'вҖҡ',
  rsquor: 'вҖҷ',
  ldquor: 'вҖһ',
  bdquo: 'вҖһ',
  comma: ',',
  period: '.',
  colon: ':',
  semi: ';',
  excl: '!',
  quest: '?',
  num: '#',
  dollar: '$',
  percent: '%',
  ast: '*',
  commat: '@',
  lowbar: '_',
  verbar: '|',
  vert: '|',
  sol: '/',
  bsol: '\\',
  lbrace: '{',
  rbrace: '}',
  lbrack: '[',
  rbrack: ']',
  lpar: '(',
  rpar: ')',
  nbsp: '\u00a0',
  iexcl: 'ВЎ',
  cent: 'Вў',
  pound: 'ВЈ',
  curren: 'ВӨ',
  yen: 'ВҘ',
  brvbar: 'ВҰ',
  sect: 'В§',
  uml: 'ВЁ',
  copy: 'В©',
  COPY: 'В©',
  ordf: 'ВӘ',
  laquo: 'В«',
  not: 'В¬',
  shy: '\u00ad',
  reg: 'В®',
  REG: 'В®',
  macr: 'ВҜ',
  deg: 'В°',
  plusmn: 'Вұ',
  sup2: 'ВІ',
  sup3: 'Ві',
  acute: 'Вҙ',
  micro: 'Вө',
  para: 'В¶',
  middot: 'В·',
  cedil: 'Вё',
  sup1: 'В№',
  ordm: 'Вә',
  raquo: 'В»',
  frac14: 'Вј',
  frac12: 'ВҪ',
  half: 'ВҪ',
  frac34: 'Вҫ',
  iquest: 'Вҝ',
  times: 'Г—',
  div: 'Г·',
  divide: 'Г·',
};

/**
 * Latin Extended & Accented Letters (A-Z)
 * @type {Record<string, string>}
 */
export const LATIN_ACCENTS = {
  Agrave: 'ГҖ',
  agrave: 'Г ',
  Aacute: 'ГҒ',
  aacute: 'ГЎ',
  Acirc: 'ГӮ',
  acirc: 'Гў',
  Atilde: 'Гғ',
  atilde: 'ГЈ',
  Auml: 'Г„',
  auml: 'ГӨ',
  Aring: 'Г…',
  aring: 'ГҘ',
  AElig: 'ГҶ',
  aelig: 'ГҰ',
  Ccedil: 'ГҮ',
  ccedil: 'Г§',
  Egrave: 'ГҲ',
  egrave: 'ГЁ',
  Eacute: 'Гү',
  eacute: 'Г©',
  Ecirc: 'ГҠ',
  ecirc: 'ГӘ',
  Euml: 'ГӢ',
  euml: 'Г«',
  Igrave: 'ГҢ',
  igrave: 'Г¬',
  Iacute: 'ГҚ',
  iacute: 'Гӯ',
  Icirc: 'ГҺ',
  icirc: 'Г®',
  Iuml: 'ГҸ',
  iuml: 'ГҜ',
  ETH: 'Гҗ',
  eth: 'Г°',
  Ntilde: 'Г‘',
  ntilde: 'Гұ',
  Ograve: 'Г’',
  ograve: 'ГІ',
  Oacute: 'Г“',
  oacute: 'Гі',
  Ocirc: 'Г”',
  ocirc: 'Гҙ',
  Otilde: 'Г•',
  otilde: 'Гө',
  Ouml: 'Г–',
  ouml: 'Г¶',
  Oslash: 'Гҳ',
  oslash: 'Гё',
  Ugrave: 'Гҷ',
  ugrave: 'Г№',
  Uacute: 'Гҡ',
  uacute: 'Гә',
  Ucirc: 'Гӣ',
  ucirc: 'Г»',
  Uuml: 'Гң',
  uuml: 'Гј',
  Yacute: 'Гқ',
  yacute: 'ГҪ',
  THORN: 'Гһ',
  thorn: 'Гҫ',
  szlig: 'Гҹ',
  yuml: 'Гҝ',
  Yuml: 'Её',
};

/**
 * Latin Extended (Letters with diacritics)
 * @type {Record<string, string>}
 */
export const LATIN_EXTENDED = {
  Amacr: 'ДҖ',
  amacr: 'ДҒ',
  Abreve: 'ДӮ',
  abreve: 'Дғ',
  Aogon: 'Д„',
  aogon: 'Д…',
  Cacute: 'ДҶ',
  cacute: 'ДҮ',
  Ccirc: 'ДҲ',
  ccirc: 'Дү',
  Cdot: 'ДҠ',
  cdot: 'ДӢ',
  Ccaron: 'ДҢ',
  ccaron: 'ДҚ',
  Dcaron: 'ДҺ',
  dcaron: 'ДҸ',
  Dstrok: 'Дҗ',
  dstrok: 'Д‘',
  Emacr: 'Д’',
  emacr: 'Д“',
  Ecaron: 'Дҡ',
  ecaron: 'Дӣ',
  Edot: 'Д–',
  edot: 'Д—',
  Eogon: 'Дҳ',
  eogon: 'Дҷ',
  Gcirc: 'Дң',
  gcirc: 'Дқ',
  Gbreve: 'Дһ',
  gbreve: 'Дҹ',
  Gdot: 'Д ',
  gdot: 'ДЎ',
  Gcedil: 'Дў',
  Hcirc: 'ДӨ',
  hcirc: 'ДҘ',
  Hstrok: 'ДҰ',
  hstrok: 'Д§',
  Itilde: 'ДЁ',
  itilde: 'Д©',
  Imacr: 'ДӘ',
  imacr: 'Д«',
  Iogon: 'Д®',
  iogon: 'ДҜ',
  Idot: 'Д°',
  IJlig: 'ДІ',
  ijlig: 'Ді',
  Jcirc: 'Дҙ',
  jcirc: 'Дө',
  Kcedil: 'Д¶',
  kcedil: 'Д·',
  kgreen: 'Дё',
  Lacute: 'Д№',
  lacute: 'Дә',
  Lcedil: 'Д»',
  lcedil: 'Дј',
  Lcaron: 'ДҪ',
  lcaron: 'Дҫ',
  Lmidot: 'Дҝ',
  lmidot: 'ЕҖ',
  Lstrok: 'ЕҒ',
  lstrok: 'ЕӮ',
  Nacute: 'Еғ',
  nacute: 'Е„',
  Ncaron: 'ЕҮ',
  ncaron: 'ЕҲ',
  Ncedil: 'Е…',
  ncedil: 'ЕҶ',
  ENG: 'ЕҠ',
  eng: 'ЕӢ',
  Omacr: 'ЕҢ',
  omacr: 'ЕҚ',
  Odblac: 'Еҗ',
  odblac: 'Е‘',
  OElig: 'Е’',
  oelig: 'Е“',
  Racute: 'Е”',
  racute: 'Е•',
  Rcaron: 'Еҳ',
  rcaron: 'Еҷ',
  Rcedil: 'Е–',
  rcedil: 'Е—',
  Sacute: 'Еҡ',
  sacute: 'Еӣ',
  Scirc: 'Ең',
  scirc: 'Еқ',
  Scedil: 'Еһ',
  scedil: 'Еҹ',
  Scaron: 'Е ',
  scaron: 'ЕЎ',
  Tcedil: 'Еў',
  tcedil: 'ЕЈ',
  Tcaron: 'ЕӨ',
  tcaron: 'ЕҘ',
  Tstrok: 'ЕҰ',
  tstrok: 'Е§',
  Utilde: 'ЕЁ',
  utilde: 'Е©',
  Umacr: 'ЕӘ',
  umacr: 'Е«',
  Ubreve: 'Е¬',
  ubreve: 'Еӯ',
  Uring: 'Е®',
  uring: 'ЕҜ',
  Udblac: 'Е°',
  udblac: 'Еұ',
  Uogon: 'ЕІ',
  uogon: 'Еі',
  Wcirc: 'Еҙ',
  wcirc: 'Еө',
  Ycirc: 'Е¶',
  ycirc: 'Е·',
  Zacute: 'Е№',
  zacute: 'Еә',
  Zdot: 'Е»',
  zdot: 'Еј',
  Zcaron: 'ЕҪ',
  zcaron: 'Еҫ',
};

/**
 * Greek Letters
 * @type {Record<string, string>}
 */
export const GREEK = {
  Alpha: 'О‘',
  alpha: 'Оұ',
  Beta: 'О’',
  beta: 'ОІ',
  Gamma: 'О“',
  gamma: 'Оі',
  Delta: 'О”',
  delta: 'Оҙ',
  Epsilon: 'О•',
  epsilon: 'Оө',
  epsiv: 'Пө',
  varepsilon: 'Пө',
  Zeta: 'О–',
  zeta: 'О¶',
  Eta: 'О—',
  eta: 'О·',
  Theta: 'Оҳ',
  theta: 'Оё',
  thetasym: 'П‘',
  vartheta: 'П‘',
  Iota: 'Оҷ',
  iota: 'О№',
  Kappa: 'Оҡ',
  kappa: 'Оә',
  kappav: 'П°',
  varkappa: 'П°',
  Lambda: 'Оӣ',
  lambda: 'О»',
  Mu: 'Оң',
  mu: 'Ој',
  Nu: 'Оқ',
  nu: 'ОҪ',
  Xi: 'Оһ',
  xi: 'Оҫ',
  Omicron: 'Оҹ',
  omicron: 'Оҝ',
  Pi: 'О ',
  pi: 'ПҖ',
  piv: 'П–',
  varpi: 'П–',
  Rho: 'ОЎ',
  rho: 'ПҒ',
  rhov: 'Пұ',
  varrho: 'Пұ',
  Sigma: 'ОЈ',
  sigma: 'Пғ',
  sigmaf: 'ПӮ',
  sigmav: 'ПӮ',
  varsigma: 'ПӮ',
  Tau: 'ОӨ',
  tau: 'П„',
  Upsilon: 'ОҘ',
  upsilon: 'П…',
  upsi: 'П…',
  Upsi: 'П’',
  upsih: 'П’',
  Phi: 'ОҰ',
  phi: 'ПҶ',
  phiv: 'П•',
  varphi: 'П•',
  Chi: 'О§',
  chi: 'ПҮ',
  Psi: 'ОЁ',
  psi: 'ПҲ',
  Omega: 'О©',
  omega: 'Пү',
  ohm: 'О©',
  Gammad: 'Пң',
  gammad: 'Пқ',
  digamma: 'Пқ',
};

/**
 * Cyrillic Letters
 * @type {Record<string, string>}
 */
export const CYRILLIC = {
  Afr: 'рқ”„',
  afr: 'рқ”һ',
  Acy: 'Рҗ',
  acy: 'Р°',
  Bcy: 'Р‘',
  bcy: 'Рұ',
  Vcy: 'Р’',
  vcy: 'РІ',
  Gcy: 'Р“',
  gcy: 'Рі',
  Dcy: 'Р”',
  dcy: 'Рҙ',
  IEcy: 'Р•',
  iecy: 'Рө',
  IOcy: 'РҒ',
  iocy: 'С‘',
  ZHcy: 'Р–',
  zhcy: 'Р¶',
  Zcy: 'Р—',
  zcy: 'Р·',
  Icy: 'Рҳ',
  icy: 'Рё',
  Jcy: 'Рҷ',
  jcy: 'Р№',
  Kcy: 'Рҡ',
  kcy: 'Рә',
  Lcy: 'Рӣ',
  lcy: 'Р»',
  Mcy: 'Рң',
  mcy: 'Рј',
  Ncy: 'Рқ',
  ncy: 'РҪ',
  Ocy: 'Рһ',
  ocy: 'Рҫ',
  Pcy: 'Рҹ',
  pcy: 'Рҝ',
  Rcy: 'Р ',
  rcy: 'СҖ',
  Scy: 'РЎ',
  scy: 'СҒ',
  Tcy: 'Рў',
  tcy: 'СӮ',
  Ucy: 'РЈ',
  ucy: 'Сғ',
  Fcy: 'РӨ',
  fcy: 'С„',
  KHcy: 'РҘ',
  khcy: 'С…',
  TScy: 'РҰ',
  tscy: 'СҶ',
  CHcy: 'Р§',
  chcy: 'СҮ',
  SHcy: 'РЁ',
  shcy: 'СҲ',
  SHCHcy: 'Р©',
  shchcy: 'Сү',
  HARDcy: 'РӘ',
  hardcy: 'СҠ',
  Ycy: 'Р«',
  ycy: 'СӢ',
  SOFTcy: 'Р¬',
  softcy: 'СҢ',
  Ecy: 'Рӯ',
  ecy: 'СҚ',
  YUcy: 'Р®',
  yucy: 'СҺ',
  YAcy: 'РҜ',
  yacy: 'СҸ',
  DJcy: 'РӮ',
  djcy: 'С’',
  GJcy: 'Рғ',
  gjcy: 'С“',
  Jukcy: 'Р„',
  jukcy: 'С”',
  DScy: 'Р…',
  dscy: 'С•',
  Iukcy: 'РҶ',
  iukcy: 'С–',
  YIcy: 'РҮ',
  yicy: 'С—',
  Jsercy: 'РҲ',
  jsercy: 'Сҳ',
  LJcy: 'Рү',
  ljcy: 'Сҷ',
  NJcy: 'РҠ',
  njcy: 'Сҡ',
  TSHcy: 'РӢ',
  tshcy: 'Сӣ',
  KJcy: 'РҢ',
  kjcy: 'Сң',
  Ubrcy: 'РҺ',
  ubrcy: 'Сһ',
  DZcy: 'РҸ',
  dzcy: 'Сҹ',
};

/**
 * Mathematical Operators & Relations
 * @type {Record<string, string>}
 */
export const MATH = {
  plus: '+',
  pm: 'Вұ',
  times: 'Г—',
  div: 'Г·',
  divide: 'Г·',
  sdot: 'вӢ…',
  star: 'вҳҶ',
  starf: 'вҳ…',
  bigstar: 'вҳ…',
  lowast: 'вҲ—',
  ast: '*',
  midast: '*',
  compfn: 'вҲҳ',
  smallcircle: 'вҲҳ',
  bullet: 'вҖў',
  bull: 'вҖў',
  nbsp: '\u00a0',
  hellip: 'вҖҰ',
  mldr: 'вҖҰ',
  prime: 'вҖІ',
  Prime: 'вҖі',
  tprime: 'вҖҙ',
  bprime: 'вҖө',
  backprime: 'вҖө',
  minus: 'вҲ’',
  minusd: 'вҲё',
  dotminus: 'вҲё',
  plusdo: 'вҲ”',
  dotplus: 'вҲ”',
  plusmn: 'Вұ',
  minusplus: 'вҲ“',
  mnplus: 'вҲ“',
  mp: 'вҲ“',
  setminus: 'вҲ–',
  smallsetminus: 'вҲ–',
  Backslash: 'вҲ–',
  setmn: 'вҲ–',
  ssetmn: 'вҲ–',
  lowbar: '_',
  verbar: '|',
  vert: '|',
  VerticalLine: '|',
  colon: ':',
  Colon: 'вҲ·',
  Proportion: 'вҲ·',
  ratio: 'вҲ¶',
  equals: '=',
  ne: 'вү ',
  nequiv: 'вүў',
  equiv: 'вүЎ',
  Congruent: 'вүЎ',
  sim: 'вҲј',
  thicksim: 'вҲј',
  thksim: 'вҲј',
  sime: 'вүғ',
  simeq: 'вүғ',
  TildeEqual: 'вүғ',
  asymp: 'вүҲ',
  approx: 'вүҲ',
  thickapprox: 'вүҲ',
  thkap: 'вүҲ',
  TildeTilde: 'вүҲ',
  ncong: 'вүҮ',
  cong: 'вү…',
  TildeFullEqual: 'вү…',
  asympeq: 'вүҚ',
  CupCap: 'вүҚ',
  bump: 'вүҺ',
  Bumpeq: 'вүҺ',
  HumpDownHump: 'вүҺ',
  bumpe: 'вүҸ',
  bumpeq: 'вүҸ',
  HumpEqual: 'вүҸ',
  le: 'вүӨ',
  LessEqual: 'вүӨ',
  ge: 'вүҘ',
  GreaterEqual: 'вүҘ',
  lesseqgtr: 'вӢҡ',
  lesseqqgtr: 'вӘӢ',
  greater: '>',
  less: '<',
};

/**
 * Mathematical Operators (Advanced)
 * @type {Record<string, string>}
 */
export const MATH_ADVANCED = {
  alefsym: 'в„ө',
  aleph: 'в„ө',
  beth: 'в„¶',
  gimel: 'в„·',
  daleth: 'в„ё',
  forall: 'вҲҖ',
  ForAll: 'вҲҖ',
  part: 'вҲӮ',
  PartialD: 'вҲӮ',
  exist: 'вҲғ',
  Exists: 'вҲғ',
  nexist: 'вҲ„',
  nexists: 'вҲ„',
  empty: 'вҲ…',
  emptyset: 'вҲ…',
  emptyv: 'вҲ…',
  varnothing: 'вҲ…',
  nabla: 'вҲҮ',
  Del: 'вҲҮ',
  isin: 'вҲҲ',
  isinv: 'вҲҲ',
  in: 'вҲҲ',
  Element: 'вҲҲ',
  notin: 'вҲү',
  notinva: 'вҲү',
  ni: 'вҲӢ',
  niv: 'вҲӢ',
  SuchThat: 'вҲӢ',
  ReverseElement: 'вҲӢ',
  notni: 'вҲҢ',
  notniva: 'вҲҢ',
  prod: 'вҲҸ',
  Product: 'вҲҸ',
  coprod: 'вҲҗ',
  Coproduct: 'вҲҗ',
  sum: 'вҲ‘',
  Sum: 'вҲ‘',
  minus: 'вҲ’',
  mp: 'вҲ“',
  plusdo: 'вҲ”',
  dotplus: 'вҲ”',
  setminus: 'вҲ–',
  lowast: 'вҲ—',
  radic: 'вҲҡ',
  Sqrt: 'вҲҡ',
  prop: 'вҲқ',
  propto: 'вҲқ',
  Proportional: 'вҲқ',
  varpropto: 'вҲқ',
  infin: 'вҲһ',
  infintie: 'в§қ',
  ang: 'вҲ ',
  angle: 'вҲ ',
  angmsd: 'вҲЎ',
  measuredangle: 'вҲЎ',
  angsph: 'вҲў',
  mid: 'вҲЈ',
  VerticalBar: 'вҲЈ',
  nmid: 'вҲӨ',
  nsmid: 'вҲӨ',
  npar: 'вҲҰ',
  parallel: 'вҲҘ',
  spar: 'вҲҘ',
  nparallel: 'вҲҰ',
  nspar: 'вҲҰ',
  and: 'вҲ§',
  wedge: 'вҲ§',
  or: 'вҲЁ',
  vee: 'вҲЁ',
  cap: 'вҲ©',
  cup: 'вҲӘ',
  int: 'вҲ«',
  Integral: 'вҲ«',
  conint: 'вҲ®',
  ContourIntegral: 'вҲ®',
  Conint: 'вҲҜ',
  DoubleContourIntegral: 'вҲҜ',
  Cconint: 'вҲ°',
  there4: 'вҲҙ',
  therefore: 'вҲҙ',
  Therefore: 'вҲҙ',
  becaus: 'вҲө',
  because: 'вҲө',
  Because: 'вҲө',
  ratio: 'вҲ¶',
  Proportion: 'вҲ·',
  minusd: 'вҲё',
  dotminus: 'вҲё',
  mDDot: 'вҲә',
  homtht: 'вҲ»',
  sim: 'вҲј',
  bsimg: 'вҲҪ',
  backsim: 'вҲҪ',
  ac: 'вҲҫ',
  mstpos: 'вҲҫ',
  acd: 'вҲҝ',
  VerticalTilde: 'вүҖ',
  wr: 'вүҖ',
  wreath: 'вүҖ',
  nsime: 'вү„',
  nsimeq: 'вү„',
  ncong: 'вүҮ',
  simne: 'вүҶ',
  ncongdot: 'в©ӯМё',
  ngsim: 'вүө',
  nsim: 'вүҒ',
  napprox: 'вүү',
  nap: 'вүү',
  ngeq: 'вүұ',
  nge: 'вүұ',
  nleq: 'вү°',
  nle: 'вү°',
  ngtr: 'вүҜ',
  ngt: 'вүҜ',
  nless: 'вү®',
  nlt: 'вү®',
  nprec: 'вҠҖ',
  npr: 'вҠҖ',
  nsucc: 'вҠҒ',
  nsc: 'вҠҒ',
};

/**
 * Arrows
 * @type {Record<string, string>}
 */
export const ARROWS = {
  larr: 'вҶҗ',
  leftarrow: 'вҶҗ',
  LeftArrow: 'вҶҗ',
  uarr: 'вҶ‘',
  uparrow: 'вҶ‘',
  UpArrow: 'вҶ‘',
  rarr: 'вҶ’',
  rightarrow: 'вҶ’',
  RightArrow: 'вҶ’',
  darr: 'вҶ“',
  downarrow: 'вҶ“',
  DownArrow: 'вҶ“',
  harr: 'вҶ”',
  leftrightarrow: 'вҶ”',
  LeftRightArrow: 'вҶ”',
  varr: 'вҶ•',
  updownarrow: 'вҶ•',
  UpDownArrow: 'вҶ•',
  nwarr: 'вҶ–',
  nwarrow: 'вҶ–',
  UpperLeftArrow: 'вҶ–',
  nearr: 'вҶ—',
  nearrow: 'вҶ—',
  UpperRightArrow: 'вҶ—',
  searr: 'вҶҳ',
  searrow: 'вҶҳ',
  LowerRightArrow: 'вҶҳ',
  swarr: 'вҶҷ',
  swarrow: 'вҶҷ',
  LowerLeftArrow: 'вҶҷ',
  lArr: 'вҮҗ',
  Leftarrow: 'вҮҗ',
  uArr: 'вҮ‘',
  Uparrow: 'вҮ‘',
  rArr: 'вҮ’',
  Rightarrow: 'вҮ’',
  dArr: 'вҮ“',
  Downarrow: 'вҮ“',
  hArr: 'вҮ”',
  Leftrightarrow: 'вҮ”',
  iff: 'вҮ”',
  vArr: 'вҮ•',
  Updownarrow: 'вҮ•',
  lAarr: 'вҮҡ',
  Lleftarrow: 'вҮҡ',
  rAarr: 'вҮӣ',
  Rrightarrow: 'вҮӣ',
  lrarr: 'вҮҶ',
  leftrightarrows: 'вҮҶ',
  rlarr: 'вҮ„',
  rightleftarrows: 'вҮ„',
  lrhar: 'вҮӢ',
  leftrightharpoons: 'вҮӢ',
  ReverseEquilibrium: 'вҮӢ',
  rlhar: 'вҮҢ',
  rightleftharpoons: 'вҮҢ',
  Equilibrium: 'вҮҢ',
  udarr: 'вҮ…',
  UpArrowDownArrow: 'вҮ…',
  duarr: 'вҮө',
  DownArrowUpArrow: 'вҮө',
  llarr: 'вҮҮ',
  leftleftarrows: 'вҮҮ',
  rrarr: 'вҮү',
  rightrightarrows: 'вҮү',
  ddarr: 'вҮҠ',
  downdownarrows: 'вҮҠ',
  har: 'вҶҪ',
  lhard: 'вҶҪ',
  leftharpoondown: 'вҶҪ',
  lharu: 'вҶј',
  leftharpoonup: 'вҶј',
  rhard: 'вҮҒ',
  rightharpoondown: 'вҮҒ',
  rharu: 'вҮҖ',
  rightharpoonup: 'вҮҖ',
  lsh: 'вҶ°',
  Lsh: 'вҶ°',
  rsh: 'вҶұ',
  Rsh: 'вҶұ',
  ldsh: 'вҶІ',
  rdsh: 'вҶі',
  hookleftarrow: 'вҶ©',
  hookrightarrow: 'вҶӘ',
  mapstoleft: 'вҶӨ',
  mapstoup: 'вҶҘ',
  map: 'вҶҰ',
  mapsto: 'вҶҰ',
  mapstodown: 'вҶ§',
  crarr: 'вҶө',
  nleftarrow: 'вҶҡ',
  nleftrightarrow: 'вҶ®',
  nrightarrow: 'вҶӣ',
  nrarr: 'вҶӣ',
  larrtl: 'вҶў',
  rarrtl: 'вҶЈ',
  leftarrowtail: 'вҶў',
  rightarrowtail: 'вҶЈ',
  twoheadleftarrow: 'вҶһ',
  twoheadrightarrow: 'вҶ ',
  Larr: 'вҶһ',
  Rarr: 'вҶ ',
  larrhk: 'вҶ©',
  rarrhk: 'вҶӘ',
  larrlp: 'вҶ«',
  looparrowleft: 'вҶ«',
  rarrlp: 'вҶ¬',
  looparrowright: 'вҶ¬',
  harrw: 'вҶӯ',
  leftrightsquigarrow: 'вҶӯ',
  nrarrw: 'вҶқМё',
  rarrw: 'вҶқ',
  rightsquigarrow: 'вҶқ',
  larrbfs: 'вӨҹ',
  rarrbfs: 'вӨ ',
  nvHarr: 'вӨ„',
  nvlArr: 'вӨӮ',
  nvrArr: 'вӨғ',
  larrfs: 'вӨқ',
  rarrfs: 'вӨһ',
  Map: 'вӨ…',
  larrsim: 'вҘі',
  rarrsim: 'вҘҙ',
  harrcir: 'вҘҲ',
  Uarrocir: 'вҘү',
  lurdshar: 'вҘҠ',
  ldrdhar: 'вҘ§',
  ldrushar: 'вҘӢ',
  rdldhar: 'вҘ©',
  lrhard: 'вҘӯ',
  uharr: 'вҶҫ',
  uharl: 'вҶҝ',
  dharr: 'вҮӮ',
  dharl: 'вҮғ',
  Uarr: 'вҶҹ',
  Darr: 'вҶЎ',
  zigrarr: 'вҮқ',
  nwArr: 'вҮ–',
  neArr: 'вҮ—',
  seArr: 'вҮҳ',
  swArr: 'вҮҷ',
  nharr: 'вҶ®',
  nhArr: 'вҮҺ',
  nlarr: 'вҶҡ',
  nlArr: 'вҮҚ',
  nrArr: 'вҮҸ',
  larrb: 'вҮӨ',
  LeftArrowBar: 'вҮӨ',
  rarrb: 'вҮҘ',
  RightArrowBar: 'вҮҘ',
};

/**
 * Geometric Shapes
 * @type {Record<string, string>}
 */
export const SHAPES = {
  square: 'в–Ў',
  Square: 'в–Ў',
  squ: 'в–Ў',
  squf: 'в–Ә',
  squarf: 'в–Ә',
  blacksquar: 'в–Ә',
  blacksquare: 'в–Ә',
  FilledVerySmallSquare: 'в–Ә',
  blk34: 'в–“',
  blk12: 'в–’',
  blk14: 'в–‘',
  block: 'в–Ҳ',
  srect: 'в–ӯ',
  rect: 'в–ӯ',
  sdot: 'вӢ…',
  sdotb: 'вҠЎ',
  dotsquare: 'вҠЎ',
  triangle: 'в–ө',
  tri: 'в–ө',
  trine: 'в–ө',
  utri: 'в–ө',
  triangledown: 'в–ҝ',
  dtri: 'в–ҝ',
  tridown: 'в–ҝ',
  triangleleft: 'в—ғ',
  ltri: 'в—ғ',
  triangleright: 'в–№',
  rtri: 'в–№',
  blacktriangle: 'в–ҙ',
  utrif: 'в–ҙ',
  blacktriangledown: 'в–ҫ',
  dtrif: 'в–ҫ',
  blacktriangleleft: 'в—Ӯ',
  ltrif: 'в—Ӯ',
  blacktriangleright: 'в–ё',
  rtrif: 'в–ё',
  loz: 'в—Ҡ',
  lozenge: 'в—Ҡ',
  blacklozenge: 'в§«',
  lozf: 'в§«',
  bigcirc: 'в—Ҝ',
  xcirc: 'в—Ҝ',
  circ: 'ЛҶ',
  Circle: 'в—Ӣ',
  cir: 'в—Ӣ',
  o: 'в—Ӣ',
  bullet: 'вҖў',
  bull: 'вҖў',
  hellip: 'вҖҰ',
  mldr: 'вҖҰ',
  nldr: 'вҖҘ',
  boxh: 'в”Җ',
  HorizontalLine: 'в”Җ',
  boxv: 'в”Ӯ',
  boxdr: 'в”Ң',
  boxdl: 'в”җ',
  boxur: 'в””',
  boxul: 'в”ҳ',
  boxvr: 'в”ң',
  boxvl: 'в”Ө',
  boxhd: 'в”¬',
  boxhu: 'в”ҙ',
  boxvh: 'в”ј',
  boxH: 'в•җ',
  boxV: 'в•‘',
  boxdR: 'в•’',
  boxDr: 'в•“',
  boxDR: 'в•”',
  boxDl: 'в••',
  boxdL: 'в•–',
  boxDL: 'в•—',
  boxuR: 'в•ҳ',
  boxUr: 'в•ҷ',
  boxUR: 'в•ҡ',
  boxUl: 'в•ң',
  boxuL: 'в•ӣ',
  boxUL: 'в•қ',
  boxvR: 'в•һ',
  boxVr: 'в•ҹ',
  boxVR: 'в• ',
  boxVl: 'в•ў',
  boxvL: 'в•Ў',
  boxVL: 'в•Ј',
  boxHd: 'в•Ө',
  boxhD: 'в•Ҙ',
  boxHD: 'в•Ұ',
  boxHu: 'в•§',
  boxhU: 'в•Ё',
  boxHU: 'в•©',
  boxvH: 'в•Ә',
  boxVh: 'в•«',
  boxVH: 'в•¬',
};

/**
 * Punctuation & Diacritics
 * @type {Record<string, string>}
 */
export const PUNCTUATION = {
  excl: '!',
  iexcl: 'ВЎ',
  brvbar: 'ВҰ',
  sect: 'В§',
  uml: 'ВЁ',
  copy: 'В©',
  ordf: 'ВӘ',
  laquo: 'В«',
  not: 'В¬',
  shy: '\u00ad',
  reg: 'В®',
  macr: 'ВҜ',
  deg: 'В°',
  plusmn: 'Вұ',
  sup2: 'ВІ',
  sup3: 'Ві',
  acute: 'Вҙ',
  micro: 'Вө',
  para: 'В¶',
  middot: 'В·',
  cedil: 'Вё',
  sup1: 'В№',
  ordm: 'Вә',
  raquo: 'В»',
  frac14: 'Вј',
  frac12: 'ВҪ',
  frac34: 'Вҫ',
  iquest: 'Вҝ',
  nbsp: '\u00a0',
  comma: ',',
  period: '.',
  colon: ':',
  semi: ';',
  vert: '|',
  Verbar: 'вҖ–',
  verbar: '|',
  dblac: 'Лқ',
  circ: 'ЛҶ',
  caron: 'ЛҮ',
  breve: 'Лҳ',
  dot: 'Лҷ',
  ring: 'Лҡ',
  ogon: 'Лӣ',
  tilde: 'Лң',
  DiacriticalGrave: '`',
  DiacriticalAcute: 'Вҙ',
  DiacriticalTilde: 'Лң',
  DiacriticalDot: 'Лҷ',
  DiacriticalDoubleAcute: 'Лқ',
  grave: '`',
};

/**
 * Currency Symbols
 * @type {Record<string, string>}
 */
export const CURRENCY = {
  cent: 'Вў',
  pound: 'ВЈ',
  curren: 'ВӨ',
  yen: 'ВҘ',
  euro: 'вӮ¬',
  dollar: '$',
  fnof: 'Ж’',
  inr: 'вӮ№',
  af: 'ШӢ',
  birr: 'бүҘбҲӯ',
  peso: 'вӮұ',
  rub: 'вӮҪ',
  won: 'вӮ©',
  yuan: 'ВҘ',
  cedil: 'Вё',
};

/**
 * Fractions
 * @type {Record<string, string>}
 */
export const FRACTIONS = {
  frac12: 'ВҪ',
  half: 'ВҪ',
  frac13: 'в…“',
  frac14: 'Вј',
  frac15: 'в…•',
  frac16: 'в…ҷ',
  frac18: 'в…ӣ',
  frac23: 'в…”',
  frac25: 'в…–',
  frac34: 'Вҫ',
  frac35: 'в…—',
  frac38: 'в…ң',
  frac45: 'в…ҳ',
  frac56: 'в…ҡ',
  frac58: 'в…қ',
  frac78: 'в…һ',
  frasl: 'вҒ„',
};

/**
 * Miscellaneous Symbols
 * @type {Record<string, string>}
 */
export const MISC_SYMBOLS = {
  trade: 'в„ў',
  TRADE: 'в„ў',
  telrec: 'вҢ•',
  target: 'вҢ–',
  ulcorn: 'вҢң',
  ulcorner: 'вҢң',
  urcorn: 'вҢқ',
  urcorner: 'вҢқ',
  dlcorn: 'вҢһ',
  llcorner: 'вҢһ',
  drcorn: 'вҢҹ',
  lrcorner: 'вҢҹ',
  intercal: 'вҠә',
  intcal: 'вҠә',
  oplus: 'вҠ•',
  CirclePlus: 'вҠ•',
  ominus: 'вҠ–',
  CircleMinus: 'вҠ–',
  otimes: 'вҠ—',
  CircleTimes: 'вҠ—',
  osol: 'вҠҳ',
  odot: 'вҠҷ',
  CircleDot: 'вҠҷ',
  oast: 'вҠӣ',
  circledast: 'вҠӣ',
  odash: 'вҠқ',
  circleddash: 'вҠқ',
  ocirc: 'вҠҡ',
  circledcirc: 'вҠҡ',
  boxplus: 'вҠһ',
  plusb: 'вҠһ',
  boxminus: 'вҠҹ',
  minusb: 'вҠҹ',
  boxtimes: 'вҠ ',
  timesb: 'вҠ ',
  boxdot: 'вҠЎ',
  sdotb: 'вҠЎ',
  veebar: 'вҠ»',
  vee: 'вҲЁ',
  barvee: 'вҠҪ',
  and: 'вҲ§',
  wedge: 'вҲ§',
  Cap: 'вӢ’',
  Cup: 'вӢ“',
  Fork: 'вӢ”',
  pitchfork: 'вӢ”',
  epar: 'вӢ•',
  ltlarr: 'вҘ¶',
  nvap: 'вүҚвғ’',
  nvsim: 'вҲјвғ’',
  nvge: 'вүҘвғ’',
  nvle: 'вүӨвғ’',
  nvlt: '<вғ’',
  nvgt: '>вғ’',
  nvltrie: 'вҠҙвғ’',
  nvrtrie: 'вҠөвғ’',
  Vdash: 'вҠ©',
  dashv: 'вҠЈ',
  vDash: 'вҠЁ',
  Vvdash: 'вҠӘ',
  nvdash: 'вҠ¬',
  nvDash: 'вҠӯ',
  nVdash: 'вҠ®',
  nVDash: 'вҠҜ',
};

export const XML = {
  amp: "&",
  apos: "'",
  gt: ">",
  lt: "<",
  quot: "\""
}
export const COMMON_HTML = {
  nbsp: '\u00a0',
  copy: '\u00a9',
  reg: '\u00ae',
  trade: '\u2122',
  mdash: '\u2014',
  ndash: '\u2013',
  hellip: '\u2026',
  laquo: '\u00ab',
  raquo: '\u00bb',
  lsquo: '\u2018',
  rsquo: '\u2019',
  ldquo: '\u201c',
  rdquo: '\u201d',
  bull: '\u2022',
  para: '\u00b6',
  sect: '\u00a7',
  deg: '\u00b0',
  frac12: '\u00bd',
  frac14: '\u00bc',
  frac34: '\u00be',
}
// ---------------------------------------------------------------------------
// Note: NUMERIC_ENTITIES (&#NNN; / &#xHH;) are handled by the scanner directly
// via String.fromCodePoint() without any map lookup.
// ---------------------------------------------------------------------------
